chalk-parse/src/ast.rs

```diff
@@ -54,6 +54,8 @@ pub struct TraitFlags {
 pub struct AssocTyDefn {
     pub name: Identifier,
     pub parameter_kinds: Vec<ParameterKind>,
+    pub bound: Option<TraitRef>,
```

Comment:
We would need multiple bounds there (see below), and using `TraitRef` is too rigid as we cannot parse things like `type Foo<T>: Iterator<Item = T>`.

Reply:
I'm thinking this would maybe need some slight refactoring for how we handle where clauses in the grammar. Something which may work would be to split existing where clauses like `TraitRef` into two parts: a left part (which would be `T` in `T: Foo<K>`) and a right part (which would be `Foo<K>`), separated by the `:` symbol. This would prevent code duplication I think because in the case of associated types, you would only ask for the right part and then build the whole `TraitRef` by hand as you did. It would be especially useful when we add lifetime requirements on types as well.

Reply:
Yep, what do you think of something like `TraitBound` and `ProjectionEqBound` that I added to ast? (Currently they're only being used for GATs.)

These can contain methods to convert them to `TraitRef`, `WhereClause::ProjectionEq`, etc. if the proper information is obtained.